Policy: Add a new ToGroups option

eloycoto opened the pull request:
Hi!

Big PR, but I tried to keep all commits small as as possible, also the
functions are super small, so should be easier to review.

Around the feature added, is the option to retrieve aws instance IPs and mark them
as allowed in a new child policy.

The overall workflow looks like this:

A new policy is added by the user, when operator receives a new event will check
if the policy has a "childrenPolicy", if the operator does not have a children
nothing will happens.

If the policy has a children policy rule, a new policy will be created, in this
case, the information will be retrieved from the policy provider information,
and AWS will return a bunch of IPs that are whitelisted, a new child policy will
be push to kubernetes by the operator and Cilium will see a CIDR policy.

In case that the information can't be retrieved, the parent policy status will
be updated with the error message.

Also, each five minutes a new controller will update the filter information.

The policy example looks like this:
```
---
kind: CiliumNetworkPolicy
apiVersion: cilium.io/v2
metadata:
  name: to-groups-sample
  namespace: default
spec:
  endpointSelector:
    matchLabels:
      any:id: app1
  egress:
  - toPorts:
    - ports:
      - port: '80'
        protocol: TCP
    toGroups:
    - aws:
        securityGroupsIds:
        - 'sg-0918ffdacf12b008a'
```

And the status in the cnp will be something like this:

```
vagrant@k8s1:~$ kubectl get cnp
NAME                                                             AGE
to-groups-sample                                                 42m
to-groups-sample-togroups-f920a466-e682-11e8-bc7e-0800271bbcb9   31m
```

The options for ToGroups are:

- SecurityGroupsNames
- SecurityGroupsIds
- Labels map[string]string
- region

Pending things to add in a future PR:

- GSG
- More unitesting. (Waiting for feedback around this)
- Add the default region from the AWS metadata url. 

If you think that is a bit PR, I'm happy to make a walk-through around the code
and explain step by step, but should be self-explained.

Regards

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6224)
<!-- Reviewable:end -->

eloycoto committed:
Policy: move ipToRules to policy/api

Due IPtoRules can be shared across multiple packages move to policy and
make it public to be used from different packages

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
Policy: Added ToGroups option into Egress policy.

To be able to add new integrations, a new field ToGroups was added.
These information will be saved in the parent rule, and if the ToGroups
is in place it'll return a new CIDR rule that will get the information
from the provider callbacks.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
ToGroups: Added initial implementation

This commit adds the initial implementation of the actions that need to
run when a new policy is added and is received by a Kubernetes watcher.

The main structure is the following:

- log: will only have the log initialiser.
- actions: the functions that are going to be triggered by the k8sevents
- helpers: some function that are used by multiple actions or
controllers.
- controllers definition of the given actions.
- Cache: is the number of policies that are now CNP that has any child
rule

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
ToGroups: Added AWS implementation

Added AWS provider implementation

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
Deps: Update AWS dependencies

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
Operator: Add CNP watcher.

To enable toGroups child policies enable watcher on the operator to add
actions when a new policy is added, deleted or updated.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
ToGroups: Add controller in operator CNP events.

Add a controller in operator to update IPs information after five
minutes to upgrade AWS ips.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
policy: Added toGroups test.

Added the test with all the options on ToGroups.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
Policy: Add egress test

Add test for HasChildRule and generate new rule for egress policies.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
CNP: Added ToGroups into parser

Added ToGroups in the cnp parser
Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
ToGroups: Add unit tests for some functions.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

tgraf commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -67,6 +67,9 @@ type CiliumNetworkPolicy struct {
 	Nodes map[string]CiliumNetworkPolicyNodeStatus `json:"nodes,omitempty"`
+
+	// ChildrenPolicies is the Cilium Children policies status
+	ChildrenPolicies map[string]CiliumChildPolicyStatus `json:"childrenPolicies,omitempty"`
```
Let's call this `Derivatives` to make it a bit clearer that we are referring to policy rules that are derived from this rule. Please extend the comment to specify what the map key describes. If it is the name of the policy, why does it have to be a map?

tgraf commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -220,6 +246,34 @@ func (r *CiliumNetworkPolicy) GetIdentityLabels() labels.LabelArray {
 		k8sCiliumUtils.ResourceTypeCiliumNetworkPolicy)
 }
 
+func (r *CiliumNetworkPolicy) HasChildrenPolicies() bool {
```
`HasDerivatives()`...

tgraf commented on pkg/policy/api/egress.go:
```diff
@@ -148,3 +159,31 @@ func (e *EgressRule) GetDestinationEndpointSelectors() EndpointSelectorSlice {
+
+// HasChildRule returns true if the policy has any rule that will create
+// a new child policy in the future.
+func (e *EgressRule) HasChildRule() bool {
```
Let's call this `RequiresDerivative()`

tgraf commented on pkg/policy/api/groups.go:
```diff
@@ -0,0 +1,91 @@
+)
+
+const (
+	AWSPROVIDER = "AWS" // AWS provider key
```
Let's not use all caps for names

tgraf commented on pkg/policy/api/rule.go:
```diff
@@ -61,3 +61,30 @@ type Rule struct {
+func (r *Rule) CreateChildRule() (*Rule, error) {
+	newRule := r.DeepCopy()
+	newRule.Egress = []EgressRule{}
+	newRule.Ingress = []IngressRule{}
```
This is definitely incorrect, it will zero out the ingress rule section without recreating it.

tgraf commented on pkg/policy/api/egress.go:
```diff
@@ -148,3 +159,31 @@ func (e *EgressRule) GetDestinationEndpointSelectors() EndpointSelectorSlice {
+
+// CreateChildRule will return a new rule based on the data gathered by
+// the rules that creates a new child policy.
+func (e *EgressRule) CreateChildRule() (*EgressRule, error) {
```
`CreateDerivative()`

tgraf commented on pkg/policy/api/groups.go:
```diff
@@ -0,0 +1,91 @@
+
+// ProviderIntegration is a func that need to be register to be able to
+// register a new provider in the platform.
+type ProviderIntegration func(*ToGroups) ([]net.IP, error)
```
Call this `GroupProviderFunc`

tgraf commented on pkg/policy/api/groups.go:
```diff
@@ -0,0 +1,91 @@
+		if !ok {
+			return emptyResult, fmt.Errorf("Provider %s is not registered", AWSPROVIDER)
+		}
+		callback := callbackInterface.(ProviderIntegration)
```
This should do `callback, ok := ...` so you can check whether the type cast was correct

tgraf commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -101,6 +104,21 @@ type CiliumNetworkPolicyNodeStatus struct {
 	Annotations map[string]string `json:"annotations,omitempty"`
 }
 
+type CiliumChildPolicyStatus struct {
```
Can you just map it to `CiliumNetworkPolicyNodeStatus`?

eloycoto replied:
No, because the node status happens on Cilium and will be overwritten.

tgraf commented on pkg/togroups/actions.go:
```diff
@@ -0,0 +1,133 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package togroups
```
This will also be used for `FromGroups` later on so a more generic name will be better.

Maybe `pkg/policy/groups/`?

tgraf commented on pkg/togroups/helpers.go:
```diff
@@ -0,0 +1,155 @@
+
+var (
+	globalK8sClient *clientset.Clientset
+	k8sMutex        = sync.Mutex{}
```
Use `pkg/lock/`

tgraf commented on pkg/togroups/actions.go:
```diff
@@ -0,0 +1,133 @@
+	var childCNP *cilium_v2.CiliumNetworkPolicy
+	var err error
+	for numAttempts := 0; numAttempts <= MaxNumberOfAttempts; numAttempts++ {
+		childCNP, err = getChildCNP(cnp)
```
What part of this can fail that justifies a retry?

eloycoto replied:
aws connection fail.

tgraf commented on pkg/togroups/actions.go:
```diff
@@ -0,0 +1,133 @@
+		return false
+	}
+
+	err = updateChildrenStatus(cnp, childCNP.ObjectMeta.Name, nil)
```
You can remove the above block and just pass `err` in

tgraf commented on pkg/togroups/aws.go:
```diff
@@ -0,0 +1,135 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package togroups
```
This should be in a separate subpackage: `pkg/policy/gropus/aws`

tgraf commented on pkg/togroups/aws.go:
```diff
@@ -0,0 +1,135 @@
+
+const (
+	awsLogLevel            = aws.LogOff // For debugging pourposes can be set to aws.LogDebugWithSigning
+	AWS_DEFAULT_REGION_KEY = "AWS_DEFAULT_REGION"
```
Please don't use all caps names

tgraf commented on pkg/togroups/aws.go:
```diff
@@ -0,0 +1,135 @@
+//given filter
+func GetInstancesIpsFromFilter(filter *api.AWSGroups) ([]net.IP, error) {
+	region := filter.Region
+	if filter.Region != "" {
```
This causes the region to be overwritten if it is already set

tgraf commented on pkg/togroups/aws.go:
```diff
@@ -0,0 +1,135 @@
+	if err != nil {
+		return nil, fmt.Errorf("Cannot initialize aws connector: %s", err)
+	}
+	cfg.Region = GetDefaultRegion()
```
Shouldn't you be using the specified region here?

tgraf commented on pkg/togroups/aws.go:
```diff
@@ -0,0 +1,135 @@
+		}
+		input.Filters = append(input.Filters, newFilter)
+	}
+	cfg, err := InitializeAWSAccount(region)
```
Do we have to do this every time we resolve? Can we do it once?

eloycoto replied:
yep :-(, each region has a different token.

tgraf commented on pkg/togroups/aws.go:
```diff
@@ -0,0 +1,135 @@
+
+//GetInstancesFromFilter returns the instances IPs in aws EC2 filter by the
+//given filter
+func GetInstancesIpsFromFilter(filter *api.AWSGroups) ([]net.IP, error) {
```
I think this will not scale well. The more CNPs you add, the more often you will have to query the metadata server. Can we fetch the entire list of IPs including labels instead and then filter on our own?

tgraf commented on pkg/togroups/aws.go:
```diff
@@ -0,0 +1,135 @@
+
+// InitializeAWSAccount retrieve the env variables from the runtime and it
+// iniliazes the account in the specified region.
+func InitializeAWSAccount(region string) (*aws.Config, error) {
```
Several functions here can be marked private

tgraf reviewed (changes requested):
@eloycoto This looks great. I've done a first pass and provided some comments. On a high level, we should avoid that importing `pkg/policy/api` pulls in the entire AWS dependencies.

ianvernon commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -67,6 +67,9 @@ type CiliumNetworkPolicy struct {
 	Nodes map[string]CiliumNetworkPolicyNodeStatus `json:"nodes,omitempty"`
+
+	// DerivativesPolicies is the Cilium Children policies status
+	DerivativesPolicies map[string]CiliumDerivativePolicyStatus `json:"DerivativesPolicies,omitempty"`
```
This should be `DerivativePolicies`.

ianvernon commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -67,6 +67,9 @@ type CiliumNetworkPolicy struct {
 	// Nodes is the Cilium policy status for each node
 	Nodes map[string]CiliumNetworkPolicyNodeStatus `json:"nodes,omitempty"`
+
+	// DerivativesPolicies is the Cilium Children policies status
```
I'd rephrase this: "the status of all policies derived from the Cilium policy"

ianvernon commented on pkg/policy/api/egress.go:
```diff
@@ -132,6 +132,17 @@ type EgressRule struct {
 	ToFQDNs []FQDNSelector `json:"toFQDNs,omitempty"`
+
+	// ToGroups is a directive that allows the integrate with multiple outside
+	// providers.  Per each field in the struct will have a register callback
```
TODO: Grammatical errors.

ianvernon commented on pkg/policy/api/egress.go:
```diff
@@ -148,3 +159,31 @@ func (e *EgressRule) GetDestinationEndpointSelectors() EndpointSelectorSlice {
+	return false
+}
+
+// CreateDerivative will return a new rule based on the data gathered by
```
I think we need to be consistent in our language. We should settle on either 'child' or 'derivative' when specifying policies which are created by other policies. I prefer 'derivative'.

ianvernon commented on pkg/policy/api/egress.go:
```diff
@@ -148,3 +159,31 @@ func (e *EgressRule) GetDestinationEndpointSelectors() EndpointSelectorSlice {
+
+// CreateDerivative will return a new rule based on the data gathered by
+// the rules that creates a new child policy.
+func (e *EgressRule) CreateDerivative() (*EgressRule, error) {
```
Add unit tests for the functions you have added in this file and throughout this commit.

eloycoto replied:
The test use data from other commit, so It's a bit hard to get all in the same commit. I can do the work, but it'll no compile.

ianvernon commented on pkg/policy/api/groups.go:
```diff
@@ -0,0 +1,94 @@
+	Aws *AWSGroups `json:"aws,omitempty"`
+}
+
+// AWSGroups is the structure that can be used to whitelisting information from AWS integration
```
TODO: grammar

ianvernon commented on pkg/policy/groups/actions.go:
```diff
@@ -0,0 +1,133 @@
+
+// AddDerivativerenCNPIfNeeded will create a new CNP if the given CNP has any rule
+// that need to create a new derivative policy.
+func AddDerivativerenCNPIfNeeded(cnp *cilium_v2.CiliumNetworkPolicy) bool {
```
Document what the boolean return value signifies.

aanm replied:
I guess derivative is typoed here as well

ianvernon commented on pkg/policy/groups/actions.go:
```diff
@@ -0,0 +1,133 @@
+	return addDerivativeCNP(cnp)
+}
+
+// UpdateDerivativeCNPIfNeeded will update or create a  CNP if the given CNP has
```
Remove space between 'a  CNP`

ianvernon commented on pkg/policy/groups/actions.go:
```diff
@@ -0,0 +1,133 @@
+}
+
+// UpdateDerivativeCNPIfNeeded will update or create a  CNP if the given CNP has
+// any rule that need to create a new derivative policy.  In case that the newCNP
```
TODO: grammar

ianvernon commented on pkg/policy/groups/actions.go:
```diff
@@ -0,0 +1,133 @@
+
+	k8sClient, err := getK8sClient()
+	if err != nil {
+		scopedLog.WithError(err).Error("Cannot get kubertenes configuration")
```
Kubernetes

ianvernon commented on pkg/policy/groups/actions.go:
```diff
@@ -0,0 +1,133 @@
+	})
+
+	if !cnp.HasDerivatives() {
+		scopedLog.Debug("CNP does not have derivatives policies, skipped")
```
grammar

ianvernon commented on pkg/policy/groups/actions.go:
```diff
@@ -0,0 +1,133 @@
+	return addDerivativeCNP(newCNP)
+}
+
+// DeleteDerivativeCNP if the given policy has any derivatives policy will be
```
grammar

raybejjani commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -220,6 +246,34 @@ func (r *CiliumNetworkPolicy) GetIdentityLabels() labels.LabelArray {
+	return false
+}
+
+func (r *CiliumNetworkPolicy) ChildrenPoliciesAlreadyEnforce() bool {
```
Since we switched to using "Derivative" this should say `DerivedPoliciesAlreadyEnforced`

raybejjani commented on pkg/policy/api/cidr.go:
```diff
@@ -133,3 +133,18 @@ func ComputeResultantCIDRSet(cidrs CIDRRuleSlice) CIDRSlice {
 }
+
+// IpsToCIDRRules generates CIDRRules for the IPs passed in.
+func IpsToCIDRRules(ips []net.IP) (cidrRules []CIDRRule) {
```
It's not a big deal, but go is ok capitalising all the letters in an abbreviation, so `IPsToCIDRRules` also works here (in fact, we already allcaps CIDR in this case)

raybejjani commented on pkg/policy/api/egress.go:
```diff
@@ -132,6 +132,17 @@ type EgressRule struct {
+	// ToGroups is a directive that allows the integrate with multiple outside
+	// providers.  Per each field in the struct will have a register callback
+	// that will return a number of IPs to be whitelisting in a new child
+	// policy.
```
Since this text shows up in the docs, it should be more high-level and oriented around how to use `toGroups`. Maybe something like:
```
ToGroups is a directive that allows integration with outside providers (e.g. AWS or GKE) of IP data. Each entry in the list corresponds to a selector specific to that provider (e.g. AWS Security Groups filtered by name or label). These rules are managed by the singleton cilium-operator and it will create derivative rules for each entry, adding toCIDRSet rules with the IPs of the group populated.
Currently only AWS is supported and the rule can select by [??].
Example:
[probably put in an actual yaml example]
```

raybejjani commented on pkg/policy/api/egress.go:
```diff
@@ -148,3 +159,31 @@ func (e *EgressRule) GetDestinationEndpointSelectors() EndpointSelectorSlice {
 }
+
+// RequiresDerivative returns true if the policy has any rule that will create
+// a new child policy in the future.
```
Since we're calling this on an EgressRule object, it might be more correct to be specific: "RequiresDerivative returns true when the EgressRule contains sections that need a derivative policy created in order to be enforced (e.g. ToGroups).

raybejjani commented on pkg/policy/api/egress.go:
```diff
@@ -148,3 +159,31 @@ func (e *EgressRule) GetDestinationEndpointSelectors() EndpointSelectorSlice {
+	if len(e.ToGroups) == 0 {
+		return newEgressRule, nil
+	}
+	newEgressRule.ToGroups = []ToGroups{}
```
Since we this always, we can skip the the early return above and leave this as nil. The loop on `ToGroups` will then be a no-op.

eloycoto replied:
The thing is, if the ToGroups is 0 the new rule will be always ToGroups{} and is not desired at all.

raybejjani replied:
Hmm, and that's because it will be serialised differently if it is nil vs an empty list? :/

aanm replied:
>The thing is, if the ToGroups is 0 the new rule will be always ToGroups{} and is not desired at all.

Why?

raybejjani commented on pkg/policy/api/egress.go:
```diff
@@ -148,3 +159,31 @@ func (e *EgressRule) GetDestinationEndpointSelectors() EndpointSelectorSlice {
+
+// CreateDerivative will return a new rule based on the data gathered by
+// the rules that creates a new child policy.
+func (e *EgressRule) CreateDerivative() (*EgressRule, error) {
```
It would be good to explain that this function will call out to get the CIDRs. In particular, where the data is being gathered and if this call is a network call or local to the operator (and maybe that this function will only work in the operator).

eloycoto replied:
done

raybejjani commented on pkg/policy/api/groups.go:
```diff
@@ -0,0 +1,105 @@
+
+var (
+	// @TODO move this to sync.Map
+	providers = providersCallbacks{}
```
was there a reason to wrap this type? From what I could see we use it as a normal sync.Map anyway, so it seems ok to just use that here.

raybejjani commented on pkg/policy/api/groups.go:
```diff
@@ -0,0 +1,105 @@
+	}
+
+	uniqueIPs := []net.IP{}
+	IPUniqueMap := map[string]bool{}
```
I doubt it matters that much in the end (performance-wise) but we can reuse this function if you prefer to only have the slice https://github.com/cilium/cilium/blob/master/pkg/fqdn/cache.go#L186-L212 In fact, you already sort them below so this is just a slight addition.

eloycoto replied:
Fixed and added another commit ;-)

raybejjani commented on pkg/policy/api/rule.go:
```diff
@@ -61,3 +61,29 @@ type Rule struct {
+// by the rules that autogenerated new Rule
+func (r *Rule) CreateDerivative() (*Rule, error) {
+	newRule := r.DeepCopy()
+	newRule.Egress = []EgressRule{}
```
The deepcopy should already have made these for us. I think this is probably fine as it is but an alternative is to keep the egressrules array here, but replace each element with the `CreateDerivative` version as you iterate (you would need to use the index). This would mean that we never touch rules that don't have a groups field. EgressRule.CreateDerivative() already does that but it isn't super clear (I guess it technically makes a copy but the outcome is the same).

eloycoto replied:
But if the copy is not in place, how can I loop over the old one? I think that is the simplest way, or I can replace at the end, but I think that it's clear on this way.

raybejjani commented on pkg/policy/groups/controllers.go:
```diff
@@ -0,0 +1,35 @@
+	for _, cnp := range cnpToUpdate {
+		sem <- true
+		go func(cnp *cilium_v2.CiliumNetworkPolicy) {
+			defer func() { <-sem }()
```
Isn't this the wrong way around? We put stuff into the sem channel, but this defer blocks the goroutine from exiting, but only after it did addDerivativeCNP and that's the thing we want to slow down.
In this case, it might be simpler to have the workers ready, and write all the CNPs in cnpToUpdate into a channel that they read from. Then, you can check a counter or something (like maybe read them all back from an output channel) to know they are ready.

eloycoto replied:
Umm, I'm not sure if I understand this one. The limit of calls to AWS is limited, so I made a function to limit to 4 in the same time,  you can slow down or you can wait until is free, is the same behaviour in a different way no?

raybejjani replied:
Oh, I understand now! yes, I see how it works. I was expecting
```
<-sem
defer func() { sem <- true }
```

at the start of each goroutine instead (and a loop to prefill the channel with `true`).  This also works but a comment  on line 29 like "This blocks when there are still maxConcurrentUpdates goroutines running" would be good.

raybejjani commented on pkg/policy/groups/helpers.go:
```diff
@@ -0,0 +1,154 @@
+)
+
+// getDerivativeCNP will return a new CNP based on the given rule.
+func getDerivativeCNP(cnp *cilium_v2.CiliumNetworkPolicy) (*cilium_v2.CiliumNetworkPolicy, error) {
```
I know we use get* in other parts of the codebase for similar functions but it's a little confusing. Since this creates an entirely new object I would prefer it to say `createDerivativeCNP` but it's not a big deal. (in contrast, getK8sClient also makes an object but then it reuses it)

raybejjani commented on pkg/policy/groups/helpers.go:
```diff
@@ -0,0 +1,154 @@
+func getK8sClient() (*clientset.Clientset, error) {
+	k8sMutex.Lock()
+	defer func() {
+		k8sMutex.Unlock()
```
This could be deferred directly, no?

raybejjani reviewed (commented):
I'll need to look at this again tomorrow, I think. I'm a little too tired to think straight. sorry!

eloycoto committed:
Policy: move KeepUniqueIPS to pkg/ip

Due this function is used in different places, make this function public
and added into pkg/ip

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

raybejjani commented on operator/cnp_event.go:
```diff
@@ -0,0 +1,91 @@
+)
+
+const (
+	reSyncPeriod = 1 * time.Second
```
I think this is the full resync time, not a poll time. We use 5 minutes in cilium-agent, and I suspect it should be much longer than 1 second here too.

raybejjani commented on pkg/policy/groups/actions.go:
```diff
@@ -0,0 +1,133 @@
+
+// DeleteDerivativeCNP if the given policy has any derivatives policy will be
+// deleted from the repo and the cache.
+func DeleteDerivativeCNP(cnp *cilium_v2.CiliumNetworkPolicy) error {
```
Does this also need to end with `IfNeeded` like Update and Add? Alternatively, we could have a helper that implements the `HasDerivatives checks` and move the IfNeeded to the place we call these functions.

raybejjani commented on pkg/policy/groups/controllers.go:
```diff
@@ -0,0 +1,35 @@
+	maxConcurrentUpdates = 4
+)
+
+func UpgradeCNPInformation() {
```
We don't really document this function, or the controller that calls it. An explanation of how it runs, and any restrictions (like only one should be running at a time) would be good. Also, is "update"  better to use here instead of upgrade?

eloycoto replied:
Fixed!

raybejjani commented on pkg/policy/groups/helpers.go:
```diff
@@ -0,0 +1,152 @@
+
+	rules, err := cnp.Parse()
+	if err != nil {
+		return nil, fmt.Errorf("Cannot parse policies: %s", err)
```
Will it be clear which parent policy we were processing? or should we include the name here?

eloycoto replied:
This information will be in the policyDerivativeStatus, so name is redundant. ;-)

raybejjani replied:
but isn't this an error message in a log somewhere?

raybejjani commented on pkg/policy/groups/helpers.go:
```diff
@@ -0,0 +1,152 @@
+	k8sCNP, err := k8sClient.CiliumV2().CiliumNetworkPolicies(cnp.ObjectMeta.Namespace).
+		Get(cnp.ObjectMeta.Name, v1.GetOptions{})
+	if err == nil {
+		cnp.ObjectMeta.ResourceVersion = k8sCNP.ObjectMeta.ResourceVersion
```
Isn't this dangerous? If the upstream CNP has annotations or something added, we would use our local copy. The local copy may be older than these new upstream additions. We may need to retry the full get-generateDerivative-update cycle if the revision number has changed.

eloycoto replied:
The CNP that we retrieve/created is a derived policy, so only need to be handle by Cilium, so will be OK.

aanm replied:
This is not safe, same CNP can be deleted and added and they will have different resource versions.

raybejjani commented on pkg/policy/groups/helpers.go:
```diff
@@ -0,0 +1,152 @@
+	}
+	k8sCNP, err := k8sClient.CiliumV2().CiliumNetworkPolicies(cnp.ObjectMeta.Namespace).
+		Get(cnp.ObjectMeta.Name, v1.GetOptions{})
+	if err == nil {
```
We may need to account for network errors here. Ideally, we would still have the network problem when we do the create below but in case of flakeyness we would see already-created errors from the `.Create` and that might be confusing to debug.

raybejjani commented on pkg/policy/groups/helpers.go:
```diff
@@ -0,0 +1,152 @@
+
+	if err != nil {
+		status.OK = false
+		status.Error = fmt.Sprintf("%v", err)
```
`err.Error()` is probably more correct here, in case the error object is something custom from k8s (or whereever).

raybejjani commented on pkg/policy/groups/helpers.go:
```diff
@@ -0,0 +1,152 @@
+	if clientErr != nil {
+		return fmt.Errorf("Cannot get Kubernetes policy: %s", clientErr)
+	}
+	if k8sCNPStatus.ObjectMeta.UID != cnp.ObjectMeta.UID {
```
Is it possible for the UID to be different? or is that the case where a new version of the CNP is uploaded. I'm not sure I understand the error case. We delete the CNP we fetched, which has a different UID from the local one. Isn't it more correct to delete the old one? (I might still need to go find the groupsCNPCache and how it works).
Relatedly, `k8sCNPStatus` was really confusing to me, since it was an actual CNP and not a status (which one can get, I think, via a getStatus call or something). upstreamCNP might be a better name but I'm not sure.

eloycoto replied:
Hi,

I didn't get this corner case, but I added that check just in case. The delete will happens on DeleteCNPIfNeeded, so no really needed, this is only a safe check. 

About delete it, I think that is worth to keep it, it'll make this easy to debug if a problem happens.

raybejjani commented on pkg/policy/groups/cache_test.go:
```diff
@@ -0,0 +1,39 @@
+	. "gopkg.in/check.v1"
+)
+
+func (s *GroupsTestSuite) TestCacheWorkingCorreclty(c *C) {
```
Correctly is typoed!

raybejjani commented on pkg/policy/groups/actions.go:
```diff
@@ -0,0 +1,133 @@
+
+	var derivativeCNP *cilium_v2.CiliumNetworkPolicy
+	var err error
+	for numAttempts := 0; numAttempts <= maxNumberOfattempts; numAttempts++ {
```
Is this loop only for retrying `createDerivativeCNP`? Is so, we could handle the error outside the loop, no? (I mean, you report the error creating it but you try again anyway. Why not try the 5 times and report that one time?). If it should also handle the `updateDerivativeStatus` call then that should also be in a loop (or the same loop).

aanm replied:
This should be a controller.

eloycoto replied:
@aanm is already called from a controller, I'm not sure what do you want here :\

raybejjani commented on pkg/policy/api/rule.go:
```diff
@@ -61,3 +61,29 @@ type Rule struct {
+	newRule.Egress = []EgressRule{}
+
+	for _, egressRule := range r.Egress {
+		devirativeEgressRule, err := egressRule.CreateDerivative()
```
derivative is typoed

raybejjani commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -151,6 +170,14 @@ func (r *CiliumNetworkPolicy) SetPolicyStatus(nodeName string, cnpns CiliumNetwo
 }
 
+// SetDerivedPolicyStatus set the status for the given children policy.
+func (r *CiliumNetworkPolicy) SetDerivedPolicyStatus(name string, status CiliumDerivativePolicyStatus) {
```
Could you change the doc comment to say that the derived policy is `name` (or change that to `derivedPolicyName`. The r here is the parent CNP, right?

raybejjani commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -220,6 +247,24 @@ func (r *CiliumNetworkPolicy) GetIdentityLabels() labels.LabelArray {
 
+// HasDerivatives return true if the CNP has any rule that will create a new
+// derivative rule.
+func (r *CiliumNetworkPolicy) HasDerivatives() bool {
```
Since the function on the Spec is `RequiresDerivative` could we make this function also `RequiresDerivative`? When I read`HasDerivative` I think it is checking that a derivative policy exists, and not that it requires one.

raybejjani commented on operator/main.go:
```diff
@@ -131,6 +131,11 @@ func runOperator(cmd *cobra.Command) {
 
+	err := EnableCNPWatcher()
+	if err != nil {
+		log.WithError(err).Fatal("Cannot connect to Kubernetes apiserver")
```
Could you make the message here a little more specific and mention the CNPwatcher? Since we seem to be talking to k8s for other things it might be confusing to find the right place

eloycoto replied:
Fixed.

raybejjani commented on pkg/policy/api/cidr.go:
```diff
@@ -133,3 +133,18 @@ func ComputeResultantCIDRSet(cidrs CIDRRuleSlice) CIDRSlice {
 	return allResultantAllowedCIDRs
 }
+
+// IPsToCIDRRules generates CIDRRules for the IPs passed in.
```
Since this is now outside fqdn the comment should mention that Generated is set to true, or we can make it a parameter.

raybejjani commented on pkg/policy/groups/actions.go:
```diff
@@ -0,0 +1,133 @@
+
+	k8sClient, err := getK8sClient()
+	if err != nil {
+		scopedLog.WithError(err).Error("Cannot get kubertenes configuration")
```
"Cannot get kubernetes client"
(kubernetes is typoed, and I changed configuration to client)

raybejjani commented on pkg/policy/groups/actions.go:
```diff
@@ -0,0 +1,133 @@
+		if statusErr != nil {
+			log.WithError(err).Error("Cannot update CNP status on invalid derivative")
+		}
+		if numAttempts == maxNumberOfattempts {
```
This should be at at the start of the loop. As it is, we will do one extra attempt before we check here.

raybejjani commented on pkg/policy/groups/aws/aws.go:
```diff
@@ -0,0 +1,135 @@
+
+// getDefaultRegion returns the given region of the default one.
+// @TODO retrieve the region from aws metadata.
+func getDefaultRegion() string {
```
Isn't this handled by the AWS client automatically?

eloycoto replied:
there is a TODO in there to get from AWS.

raybejjani replied:
wow... good job AWS :/ :\

raybejjani commented on pkg/policy/groups/aws/aws.go:
```diff
@@ -0,0 +1,135 @@
+	for _, reservation := range req.Reservations {
+		for _, instance := range reservation.Instances {
+			for _, iface := range instance.NetworkInterfaces {
+				for _, ifaceIP := range iface.PrivateIpAddresses {
```
We may also need to include public EIPs here too.

eloycoto replied:
public IPs are already in there.

raybejjani commented on pkg/policy/api/egress_test.go:
```diff
@@ -0,0 +1,85 @@
+	c.Assert(err, IsNil)
+}
+
+func (s *PolicyAPITestSuite) TestCreateDerivativeRuleWithToGroupsWitInvalidRegisterCallback(c *C) {
```
Wit -> With but a comment might be much easier to read for all these test functions.

eloycoto replied:
done.

raybejjani commented on pkg/policy/groups/actions.go:
```diff
@@ -0,0 +1,133 @@
+// any rule that need to create a new derivative policy.  In case that the newCNP
+// will not have any derivative policy and the old one had one, it'll delete the
+// old policy.
+func UpdateDerivativeCNPIfNeeded(newCNP *cilium_v2.CiliumNetworkPolicy, oldCNP *cilium_v2.CiliumNetworkPolicy) bool {
```
derivative is typoed

aanm replied:
Please update documentation, I still don't know what this function is suppose to do or not do.

eloycoto replied:
Done.

raybejjani commented on pkg/policy/groups/controllers.go:
```diff
@@ -0,0 +1,40 @@
+// from providers.  To avoid issues with rate-limiting this function will
+// execute the addDerivative function with a max number of concurrent calls,
+// defined on maxConcurrentUpdates.
+func UpgradeCNPInformation() {
```
Could you rename this to `UpdateCNPInformation` "Upgrade" sounds like we're changing a version.

eloycoto replied:
done

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -67,6 +67,9 @@ type CiliumNetworkPolicy struct {
 	Nodes map[string]CiliumNetworkPolicyNodeStatus `json:"nodes,omitempty"`
+
+	// DerivativesPolicies is the Cilium Children policies status
+	DerivativesPolicies map[string]CiliumDerivativePolicyStatus `json:"DerivativesPolicies,omitempty"`
```
`s/DerivativesPolicies/DependentPolicies`

eloycoto replied:
Hi, 

I'm not agree with the *dependant* name, also was @tgraf  who suggest the Derivative Name.

For me, derivative is something that is created from the parent (Before I called children) and create something new. That is what this code is doing.

If we call dependant, means that it's relying on something else for support, but in this case, the field is on the parent policy, so it's not a dependant at all.

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -67,6 +67,9 @@ type CiliumNetworkPolicy struct {
 	Nodes map[string]CiliumNetworkPolicyNodeStatus `json:"nodes,omitempty"`
+
+	// DerivativesPolicies is the Cilium Children policies status
+	DerivativesPolicies map[string]CiliumDerivativePolicyStatus `json:"DerivativesPolicies,omitempty"`
```
`json:"DerivativesPolicies,omitempty"` -> `json:"dependentPolicies,omitempty"`

eloycoto replied:
https://github.com/cilium/cilium/pull/6224/commits/cbd97fed812cc4893242a8264f62035d24e61053#r234515061

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -101,6 +104,22 @@ type CiliumNetworkPolicyNodeStatus struct {
 }
 
+// CiliumDerivativePolicyStatus is the status of a Cilium derived policies.
+type CiliumDerivativePolicyStatus struct {
```
`CiliumDependentPolicyStatus`

eloycoto replied:
https://github.com/cilium/cilium/pull/6224/commits/cbd97fed812cc4893242a8264f62035d24e61053#r234515061

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -101,6 +104,22 @@ type CiliumNetworkPolicyNodeStatus struct {
 
+// CiliumDerivativePolicyStatus is the status of a Cilium derived policies.
+type CiliumDerivativePolicyStatus struct {
+	// OK is true when the policy has been parsed and imported successfully
```
This is basically `CiliumNetworkPolicyNodeStatus`. As the fields are omitted if empty you can use CiliumNetworkPolicyNodeStatus directly.

eloycoto replied:
No, it's not the same. Here is the status of the children&derivative&dependant policy. Having into node does not make sense, due is the controller, not the policy who is updating it.

ianvernon replied:
@eloycoto I think @aanm is saying that the type is the same as the CiliumNetworkPolicyNodeStatus - couldn't we declare it `type CiliumDerivativePolicyStatus CiliumNetworkPolicyNodeStatus`?

eloycoto replied:
Done.

raybejjani reviewed (commented):
I keep having trouble tracking what exact CNP we are manipulating while reading the code. Would it be possible to be strict and label all parent CNPs `parentCNP` when they are a variable, derived CNPs `derivedCNP`, and when the cnp is any cnp, leave it as `cnp`?

I think I finally figured out why I feel uneasy with parts of this change. I feel we added methods to pkg/policy/api for things that are specific to groups. For example:
`EgressRule.RequiresDerivative` and `.CreateDerivative`. The meaning of derivative here is specific to groups, and the future changes we expect to add are also specific to groups (e.g. GKE support). api.EgressRule itself is more of a simple carrier of information.
More weird is that the `AWSGroups` field implements working logic related to groups (via GetCidrSet). This isn’t wrong, since that’s what `AWSGroups` there for, but it’s a little weird that pkg/policy/groups/aws registers with pkg/policy/api (via RegisterToGroupsProvider) but then the various CreateDerivative functions in pkg/policy/api are called by pkg/policy/groups. It’s like we separated the helpers to make derivatives (createDerivative) from the logic that uses them (pkg/policy/groups), and connected the source of the data (pkg/policy/groups/aws) to the object that stores the final output (pkg/policy/api).

I might have the wrong mental model. In my head `pkg/policy/api` is the most basic data as it is seen in the json, with very few methods. `pkg/policy/groups` is where we manage the data from `pkg/policy/api`. We would create a derivative, replace toGroups with toCIDRSet etc. Helpers to do all this would be here too. `pkg/policy/groups/aws` would specifically have functions to talk to aws and do actions with it (the filter call, or a watcher). It would also convert to common data types (net.IP etc.) or, maybe, directly to toCIDRSet (but, `api.IPsToCIDRRules` feels like the correct place for it, since it is a utility function to convert a common type to a specific type in the api package).
If we go with what I’m saying, I think the code structure would change a little.
`pkg/policy/api` only has the ToGroups field and AWSGroups as structs
`pkg/policy/groups` has all the functions like NeedsDerivative, CreateDerivative etc. and the updateCNP type functions, and the list of toGroups providers.
`pkg/policy/groups/aws` would be just for the AWS interaction. This will be a little annoying, since the `RegisterToGroupsProvider` call can’t be done in `pkg/policy/groups/aws`, because `pkg/policy/groups` will import `pkg/policy/groups/aws`. We can still keep the `RegisterToGroupsProvider` style if you prefer, but that would need a third package in that case.

aanm commented on pkg/policy/api/egress.go:
```diff
@@ -132,6 +132,18 @@ type EgressRule struct {
+	// toGroups:
+	// - aws:
+	//     securityGroupsIds:
+	//     - 'sg-XXXXXXXXXXXXX'
```
Create GH issue or add CRD validation in this PR

aanm commented on pkg/policy/api/egress.go:
```diff
@@ -148,3 +160,33 @@ func (e *EgressRule) GetDestinationEndpointSelectors() EndpointSelectorSlice {
+// RequiresDerivative returns true when the EgressRule contains sections that
+// need a derivative policy created in order to be enforced (e.g. ToGroups).
+func (e *EgressRule) RequiresDerivative() bool {
+	if len(e.ToGroups) > 0 {
```
`return len(e.ToGroups) > 0`

aanm commented on pkg/policy/api/groups.go:
```diff
@@ -0,0 +1,89 @@
+)
+
+var (
+	providers = sync.Map{}
```
Provide documentation for this variable

aanm commented on pkg/policy/api/groups.go:
```diff
@@ -0,0 +1,89 @@
+// ToGroups structure to store all kinds of new integrations that needs a new
+// derivative policy.
+type ToGroups struct {
+	Aws *AWSGroups `json:"aws,omitempty"`
```
`Aws` -> `AWS`

aanm commented on pkg/policy/api/groups.go:
```diff
@@ -0,0 +1,89 @@
+}
+
+// AWSGroups is the structure that can be used to whitelisting information from AWS integration
+type AWSGroups struct {
```
`AWSGroups` -> `AWSGroup`?

eloycoto replied:
for me is more than one, but changed it.

aanm commented on pkg/policy/api/groups.go:
```diff
@@ -0,0 +1,89 @@
+	if group.Aws != nil {
+		callbackInterface, ok := providers.Load(AWSProvider)
+		if !ok {
+			return emptyResult, fmt.Errorf("Provider %s is not registered", AWSProvider)
```
return nil

eloycoto replied:
No, in this case should provide a empty result, if the data from AWS can't not be retrieved the policy shouldn't allow all. (Discussed on the Oct 26th meeting)

ianvernon replied:
Hm, shouldn't we provide nil since an empty slice allows all?

eloycoto replied:
moved to nil. 

Empty slice is blockAll too, because no servers found does not mean that we allow all.

aanm commented on pkg/policy/api/groups.go:
```diff
@@ -0,0 +1,89 @@
+		}
+		callback, ok := callbackInterface.(GroupProviderFunc)
+		if !ok {
+			return emptyResult, fmt.Errorf("Provider callback for %s is not a valid instance", AWSProvider)
```
return nil

eloycoto replied:
No, in this case should provide a empty result, if the data from AWS can't not be retrieved the policy shouldn't allow all. (Discussed on the Oct 26th meeting)

ianvernon replied:
Same here.

eloycoto replied:
moved to nil.

Empty slice is blockAll too, because no servers found does not mean that we allow all.

aanm commented on pkg/policy/api/groups.go:
```diff
@@ -0,0 +1,89 @@
+		}
+		awsIPs, err := callback(group)
+		if err != nil {
+			return emptyResult, fmt.Errorf(
```
return nil

eloycoto replied:
No, in this case should provide a empty result, if the data from AWS can't not be retrieved the policy shouldn't allow all. (Discussed on the Oct 26th meeting)

ianvernon replied:
And here.

eloycoto replied:
moved to nil.

Empty slice is blockAll too, because no servers found does not mean that we allow all.

aanm commented on pkg/policy/api/rule.go:
```diff
@@ -61,3 +61,29 @@ type Rule struct {
+	for _, egressRule := range r.Egress {
+		devirativeEgressRule, err := egressRule.CreateDerivative()
+		if err != nil {
+			return newRule, err
```
`return nil, err`

eloycoto replied:
No, in this case should provide a empty result, if the data from AWS can't not be retrieved the policy shouldn't allow all. (Discussed on the Oct 26th meeting)

aanm commented on pkg/policy/api/egress.go:
```diff
@@ -148,3 +160,33 @@ func (e *EgressRule) GetDestinationEndpointSelectors() EndpointSelectorSlice {
+// In the case of ToGroups will call outside using the groups callback and this
+// function can take a bit of time.
+func (e *EgressRule) CreateDerivative() (*EgressRule, error) {
+	newEgressRule := e.DeepCopy()
```
This DeepCopy() was already executed by the caller of this function. Consider making `CreateDerivative` method private and add in the documentation stating the EgressRule will be modified in the receiver so you won't return an EgressRule. TBH I don't know what's the best approach here but we need to avoid doing 2 deepcopies cc @raybejjani

aanm commented on pkg/policy/groups/actions.go:
```diff
@@ -0,0 +1,133 @@
+// Copyright 2017-2018 Authors of Cilium
```
2018

aanm commented on pkg/policy/groups/actions.go:
```diff
@@ -0,0 +1,133 @@
+)
+
+const (
+	maxNumberOfattempts = 5
```
write documentation

aanm commented on pkg/policy/groups/actions.go:
```diff
@@ -0,0 +1,133 @@
+
+// DeleteDerivativeCNPIfNeeded if the given policy has any derivatives policy will be
+// deleted from the repo and the cache.
+func DeleteDerivativeCNPIfNeeded(cnp *cilium_v2.CiliumNetworkPolicy) error {
```
Consider making this a method of cilium network policy.

aanm replied:
Actually this function should not be needed it, I'll explain in the overview.

aanm commented on pkg/policy/groups/cache.go:
```diff
@@ -0,0 +1,44 @@
+// Copyright 2017-2018 Authors of Cilium
```
2018

aanm commented on pkg/policy/groups/controllers.go:
```diff
@@ -0,0 +1,49 @@
+// Copyright 2017-2018 Authors of Cilium
```
2018

aanm commented on pkg/policy/groups/controllers.go:
```diff
@@ -0,0 +1,49 @@
+		for _, cnp := range cnpToUpdate {
+			sem <- true
+			go func(cnp *cilium_v2.CiliumNetworkPolicy) {
+				defer func() { <-sem }()
```
User a `sync.WaitGroup` is less complicated to understand what you were trying to achive with the channels.

aanm commented on pkg/policy/groups/helpers.go:
```diff
@@ -0,0 +1,152 @@
+// Copyright 2017-2018 Authors of Cilium
```
2018

aanm commented on pkg/policy/groups/helpers.go:
```diff
@@ -0,0 +1,152 @@
+// createDerivativeCNP will return a new CNP based on the given rule.
+func createDerivativeCNP(cnp *cilium_v2.CiliumNetworkPolicy) (*cilium_v2.CiliumNetworkPolicy, error) {
+	derivativeCNP := cnp.DeepCopy()
+	derivativeCNP.ObjectMeta.Name = fmt.Sprintf(
```
1) Create function to derive name of a cnp.

2) delete the whole ObjectMeta
```
derivativeCNP = v1.ObjectMeta {
   Name: GetDependentGroupNameOf(cnp),
   Namespace: cnp.ObjectMeta.Namespace,
   Labels: map[string]string{
		parentCNP:  string(cnp.ObjectMeta.UID),
		cnpKindKey: cnpKindName,
	},
   OwnerReferences: []v1.OwnerReferences{
      APIVersion: cnp.ObjectMeta.APIVersion,
      Kind: cnp.ObjectMeta.Kind,
      UID: cnp.ObjectMeta.UID,
      Name: cnp.ObjectMeta.Name,
      Namespace: cnp.ObjectMeta.Namespace,
   },
```

aanm commented on pkg/policy/groups/helpers.go:
```diff
@@ -0,0 +1,152 @@
+
+// createDerivativeCNP will return a new CNP based on the given rule.
+func createDerivativeCNP(cnp *cilium_v2.CiliumNetworkPolicy) (*cilium_v2.CiliumNetworkPolicy, error) {
+	derivativeCNP := cnp.DeepCopy()
```
If you are deleting the Spec and Specs, why are you creating a DeepCopy at all?

eloycoto replied:
labels, and objectmeta.

aanm commented on pkg/policy/groups/helpers.go:
```diff
@@ -0,0 +1,152 @@
+	return k8sClient.CiliumV2().CiliumNetworkPolicies(cnp.ObjectMeta.Namespace).Create(cnp)
+}
+
+func updateDerivativeStatus(cnp *cilium_v2.CiliumNetworkPolicy, derivativeName string, err error) error {
```
We already have a function that updates status for a cnp with a lot of controllers, can't we use that one instead of doing new mistakes?

eloycoto replied:
could you point to me what is that function? I can't find it.

raybejjani replied:
I think aanm means this one (maybe) https://github.com/cilium/cilium/blob/master/daemon/k8s_watcher.go#L1325

aanm commented on pkg/policy/groups/aws/aws.go:
```diff
@@ -0,0 +1,135 @@
+const (
+	awsLogLevel         = aws.LogOff // For debugging pourposes can be set to aws.LogDebugWithSigning
+	awsDefaultRegionKey = "AWS_DEFAULT_REGION"
+	awsDefaultRegion    = "eu-west-1"
```
Why do we have a default region?

eloycoto replied:
Is a todo function that will happen in a follow-up PR.

aanm commented on pkg/policy/groups/aws/aws.go:
```diff
@@ -0,0 +1,135 @@
+	api.RegisterToGroupsProvider(api.AWSPROVIDER, GetIPsFromGroup)
+}
+
+// GetIpsFromGroup will return the list of the ips for the given group filter
```
this

aanm commented on pkg/policy/groups/aws/aws.go:
```diff
@@ -0,0 +1,135 @@
+)
+
+var (
+	policySecurityGroupIdKey = aws.String("instance.group-id")
```
this

aanm commented on pkg/policy/groups/aws/aws.go:
```diff
@@ -0,0 +1,135 @@
+func GetIPsFromGroup(group *api.ToGroups) ([]net.IP, error) {
+	result := []net.IP{}
+	if group.Aws == nil {
+		return result, fmt.Errorf("no aws data available")
```
`return nil, errors.New("no aws data available")`

eloycoto replied:
The same as in the past, in case of error we're interested on a empty result/CIDRSet.

aanm commented on pkg/policy/groups/aws/aws.go:
```diff
@@ -0,0 +1,135 @@
+		return result, fmt.Errorf("no aws data available")
+	}
+	ips, err := getInstancesIpsFromFilter(group.Aws)
+	return ips, err
```
`return getInstancesIpsFromFilter(group.Aws)`

aanm commented on pkg/policy/groups/aws/aws.go:
```diff
@@ -0,0 +1,135 @@
+
+func awsDumpIpsFromRequest(req *ec2.DescribeInstancesOutput) []net.IP {
+	result := []net.IP{}
+	if len(req.Reservations) == 0 {
```
you can remove these 3 lines, the for loop on line 122 is not executed if `len(req.Reservations) == 0`

aanm commented on pkg/policy/groups/providers.go:
```diff
@@ -0,0 +1,20 @@
+
+// Empty imports to register providers
+import (
+	_ "github.com/cilium/cilium/pkg/policy/groups/aws"
```
This is not resolved, add the comment right before the import

aanm commented on pkg/policy/api/groups.go:
```diff
@@ -80,10 +79,6 @@ func (group *ToGroups) GetCidrSet() ([]CIDRRule, error) {
-		return bytes.Compare(ips[i], ips[j]) < 0
-	})
-	return IPsToCIDRRules(ips), nil
+	resultIps := KeepUniqueIPs(ips)
```
What happened here?

eloycoto replied:
A new helper was used from the FQDN package.

aanm commented on pkg/policy/api/groups_test.go:
```diff
@@ -0,0 +1,113 @@
+}
+
+func (s *PolicyAPITestSuite) TestGetCIDRSetWithMultipleSorted(c *C) {
+	cb, cidrRule := GetCallBackWithRule("192.168.1.1", "192.168.10.10")
```
Do `"192.168.1.1", "192.168.10.10", "192.168.10.3"`

aanm commented on pkg/policy/api/groups_test.go:
```diff
@@ -0,0 +1,113 @@
+	group := GetToGroupsRule()
+	cidr, err := group.GetCidrSet()
+	c.Assert(cidr, DeepEquals, cidrRule)
+	c.Assert(err, IsNil)
```
Verify if they are sorted

eloycoto replied:
It's already verified by the expectedCIDRRule.

aanm commented on pkg/policy/api/egress_test.go:
```diff
@@ -0,0 +1,85 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (s *PolicyAPITestSuite) TestHasDerivativeRuleWithoutToGroups(c *C) {
```
Change the test name to `TestRequiresDerivative` and run multiple tests in it. This function and the next one should be joined.

eloycoto replied:
I do not agree with this one :-(. 

One test should test only one thing. If we have a test that is TestHasDerivative should test only *one* thing.

Larger test means that should have logic, lest factories, builders, etc.. and test description is not what the test have.

aanm commented on pkg/policy/api/egress_test.go:
```diff
@@ -0,0 +1,85 @@
+	c.Assert(eg.RequiresDerivative(), Equals, false)
+}
+
+func (s *PolicyAPITestSuite) TestCreateDerivativeRuleWithoutToGroups(c *C) {
```
Same here `TestCreateDerivative`

eloycoto replied:
I do not agree with this one :-(.

One test should test only one thing. If we have a test that is TestHasDerivative should test only one thing.

Larger test means that should have logic, lest factories, builders, etc.. and test description is not what the test have.

aanm commented on pkg/policy/groups/cache_test.go:
```diff
@@ -0,0 +1,39 @@
+// Copyright 2017-2018 Authors of Cilium
```
date

aanm commented on pkg/policy/groups/groups_test.go:
```diff
@@ -0,0 +1,32 @@
+// Copyright 2017-2018 Authors of Cilium
```
date

aanm commented on pkg/policy/groups/helpers_test.go:
```diff
@@ -0,0 +1,120 @@
+// Copyright 2017-2018 Authors of Cilium
```
date

aanm reviewed (changes requested):
Kubernetes allows to have Dependents of a particular object. If the "owner" of that object is deleted, all of the dependents are deleted as well.

```
$ kubectl get cnp fancyrule 
NAME        AGE
fancyrule   7m
$ kubectl get cnp fancyrule-2
NAME          AGE
fancyrule-2   4m
$ kubectl delete cnp fancyrule 
ciliumnetworkpolicy.cilium.io "fancyrule" deleted
$ kubectl get cnp fancyrule-2
Error from server (NotFound): ciliumnetworkpolicies.cilium.io "fancyrule-2" not found
```

This can be achieved by having the owner references in the dependent CNP
```
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  creationTimestamp: 2018-11-16T16:53:49Z
  generation: 1
  name: fancyrule-2
  namespace: default
  ownerReferences:
  - apiVersion: cilium.io/v2
    blockOwnerDeletion: true
    kind: CiliumNetworkPolicy
    name: fancyrule
    uid: ac19702a-e9bf-11e8-b8e7-0800271bbcb9
  resourceVersion: "391166"
  selfLink: /apis/cilium.io/v2/namespaces/default/ciliumnetworkpolicies/fancyrule-2
  uid: 30d26758-e9c0-11e8-b8e7-0800271bbcb9
```

"Owner" policy
```
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  creationTimestamp: 2018-11-16T16:50:07Z
  generation: 1
  name: fancyrule
  namespace: default
  resourceVersion: "391183"
  selfLink: /apis/cilium.io/v2/namespaces/default/ciliumnetworkpolicies/fancyrule
  uid: ac19702a-e9bf-11e8-b8e7-0800271bbcb9
```

eloycoto committed:
K8s: Add CNPKindDefinition const

Due is not widely used around multiple places, added a new const to be
used in different packages.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
ToGroups: Added Kubernetes validation

Added kubernetes validation for ToGroups and AWSGroups

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

ianvernon commented on pkg/policy/api/utils.go:
```diff
@@ -108,3 +111,31 @@ func ParseL4Proto(proto string) (L4Proto, error) {
+// will not overwrite a valid element with another. To overwrite an element i
+// with j, i must have come before j AND we decided it was a duplicate of the
+// element at i-1.
+func KeepUniqueIPs(ips []net.IP) []net.IP {
```
Nit - this would fit better in `pkg/ip`.

eloycoto replied:
Done!

ianvernon commented on pkg/policy/api/egress.go:
```diff
@@ -148,3 +160,37 @@ func (e *EgressRule) GetDestinationEndpointSelectors() EndpointSelectorSlice {
+			return nil, err
+		}
+		if len(cidrSet) == 0 {
+			blockAll()
```
Add a unit test for the blockAll case.

ianvernon commented on pkg/policy/api/egress.go:
```diff
@@ -148,3 +160,37 @@ func (e *EgressRule) GetDestinationEndpointSelectors() EndpointSelectorSlice {
+			return nil, err
+		}
+		if len(cidrSet) == 0 {
+			blockAll()
```
What will the status of a child policy look like in the block all case?

eloycoto replied:
I'm not sure If I understand this one. It'll be a egress block all child policy and the status will be enforce. The parent policy will have the error in the derivativeStatus.

ianvernon commented on pkg/policy/api/groups.go:
```diff
@@ -0,0 +1,89 @@
+type GroupProviderFunc func(*ToGroups) ([]net.IP, error)
+
+// ToGroups structure to store all kinds of new integrations that needs a new
+// derivative policy.
```
A lot of the documentation needs correcting / grammatical fixes. This can be done in a follow-up.

ianvernon commented on Gopkg.toml:
```diff
@@ -306,3 +306,7 @@ required = [
+
+[[constraint]]
+  name = "github.com/aws/aws-sdk-go-v2"
+  version = "2.0.0-preview.5"
```
Is it possible to use a different dependency version that is not a 'preview'? I'm wary of using dependencies that aren't official releases.

eloycoto replied:
AWS has a new SDK, that it's easy to use and is in place for near a year, but the versions are not still stable. For our use case works ok. 

We can use the old sdk, but I prefer to use the new one to not change in the near future. 

You can read more about here: 
https://aws.amazon.com/blogs/developer/aws-sdk-for-go-2-0-developer-preview/

ianvernon commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -540,9 +540,39 @@ var (
 				},
 			},
+			"toGroups": {
+				Description: `ToGroups is a list of contraninst that will
```
constraints

ianvernon commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -540,9 +540,39 @@ var (
 			},
+			"toGroups": {
+				Description: `ToGroups is a list of contraninst that will
+				gather data from third party provieders and create a new
```
providers

ianvernon replied:
third-party

ianvernon commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -540,9 +540,39 @@ var (
+		Description: "",
+		Properties: map[string]apiextensionsv1beta1.JSONSchemaProps{
+			"SecurityGroupsIds": {
+				Description: `SecurityGroupsIds is the list of Amazon aws
```
Amazon aws --> AWS

ianvernon commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -540,9 +540,39 @@ var (
+				Type: "array",
+			},
+			"SecurityGroupsNames": {
+				Description: `SecurityGroupsNames is the list of Amazon aws
```
Amazon aws --> AWS

ianvernon commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -540,9 +540,39 @@ var (
+		Properties: map[string]apiextensionsv1beta1.JSONSchemaProps{
+			"SecurityGroupsIds": {
+				Description: `SecurityGroupsIds is the list of Amazon aws
+				security groups IDs that will filter the instances IPs from AWS
```
security group IDs

ianvernon replied:
from the AWS API

ianvernon commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -540,9 +540,39 @@ var (
+			},
+			"SecurityGroupsNames": {
+				Description: `SecurityGroupsNames is the list of Amazon aws
+				security groups names that will filter the instances IPs from
```
from the AWS API

ianvernon commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -540,9 +540,39 @@ var (
+				Type: "array",
+			},
+			"Region": {
+				Description: `Region is the key that will filter the aws EC2
```
aws --> AWS

aanm commented on daemon/k8s_watcher.go:
```diff
@@ -561,7 +561,11 @@ func (d *Daemon) EnableK8sWatcher(reSyncPeriod time.Duration) error {
-					err := d.addCiliumNetworkPolicyV2(cnpStore, i.(*cilium_v2.CiliumNetworkPolicy))
+					cnp := i.(*cilium_v2.CiliumNetworkPolicy)
+					if cnp.RequiresDerivative() {
+						return nil
```
How is the parent CNP update its status?

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -67,6 +67,10 @@ type CiliumNetworkPolicy struct {
+
+	// DerivativePolicies is the status of all policies derived from the Cilium
+	// policy
+	DerivativePolicies map[string]CiliumNetworkPolicyNodeStatus `json:"DerivativePolicies,omitempty"`
```
`json:"derivativePolicies,omitempty"`

eloycoto committed:
ToGroups: Fix issues when ToGroups fail.

If the policy fails, the egressrule default block was not installed at
all. It works when the result is a empty slice.

With this change, in case of a fail connection the egressrule default
all will alwauys be installed.

Also the following changes were made:
- Update CreateDerivative egress to make a deepcopy.
- Update CreateDerivative helper to create a new rule instead of
deepcopy the given CNP.
- Fix error message on status update.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

aanm commented on pkg/policy/api/groups.go:
```diff
@@ -0,0 +1,87 @@
+)
+
+var (
+	providers = sync.Map{} // map with the list of providers to callback to retrive info from.
```
This does not need to be a sync.Map because we do the provier registration at init(). It's better to switch to a simple map by now than have the cost of doing a `Load` every time we want to load the policy.

eloycoto replied:
The reason to have a sync.Map is because of the reads on GetCidrSet()

aanm commented on pkg/policy/api/groups.go:
```diff
@@ -0,0 +1,87 @@
+				"Cannot retrieve data from %s provider: %s",
+				AWSProvider, err)
+		}
+		ips = append(ips, awsIPs...)
```
This doesn't need to be an `append`, we can initialize the `awsIPs` variable on line 61.

eloycoto replied:
the idea is to have more providers in the future.

aanm commented on pkg/policy/groups/actions.go:
```diff
@@ -0,0 +1,169 @@
+		}).Debug("CNP does not have derivative policies, skipped")
+		return true
+	}
+	controllerManager.UpdateController(fmt.Sprintf("CNP-Derivative-add-%s", cnp.ObjectMeta.Name),
```
`"CNP-Derivative-add-%s"` -> `"add-derivative-cnp-%s"`

aanm commented on pkg/policy/groups/actions.go:
```diff
@@ -0,0 +1,169 @@
+			logfields.K8sNamespace:            newCNP.ObjectMeta.Namespace,
+		}).Info("New CNP does not have derivative policy, but old had. Deleted old policies")
+
+		controllerManager.UpdateController(fmt.Sprintf("CNP-Derivative-delete-%s", oldCNP.ObjectMeta.Name),
```
`"CNP-Derivative-delete-%s"` -> `"add-derivative-cnp-%s"`

aanm commented on pkg/policy/groups/controllers.go:
```diff
@@ -0,0 +1,51 @@
+	maxConcurrentUpdates = 4
+)
+
+func init() {
```
Won't this code will be shared between cilium operator and cilium agent? Do we want the `init()` to be called for cilium-agent?

aanm commented on pkg/policy/groups/controllers.go:
```diff
@@ -0,0 +1,51 @@
+// from providers.  To avoid issues with rate-limiting this function will
+// execute the addDerivative function with a max number of concurrent calls,
+// defined on maxConcurrentUpdates.
+func UpdateCNPInformation() {
```
Why aren't we using a watcher instead? Can be done as a follow up. I think I don't understand entirely what this is doing.

eloycoto replied:
No, this fetch info from AWS each 5 minutes and create the new policy. The whatcher in AWS means that some changes need to be done in the user AWS account. (CloudWatch+Lambda applications)

aanm commented on operator/cnp_event.go:
```diff
@@ -80,5 +81,14 @@ func enableCNPWatcher() error {
 	ciliumV2Controller.AddEventHandler(watcher)
 	si.Start(wait.NeverStop)
 
+	controller.NewManager().UpdateController("cnp-to-groups-controller",
```
`"cnp-to-groups-controller"` -> `"cnp-to-groups"`

aanm commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -540,9 +540,37 @@ var (
 		},
 	}
 
+	AWSGroup = apiextensionsv1beta1.JSONSchemaProps{
```
Bump the number of the CRD validation version. (It should be on the top of this file)

aanm commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -540,9 +540,37 @@ var (
+			"SecurityGroupsIds": {
+				Description: `SecurityGroupsIds is the list of AWS security
+				group IDs that will filter the instances IPs from the AWS API`,
+				Type: "array",
```
Can be done as a follow up but I would like to see the validation done for the `groupIDs` object. There must be a defined regex in aws docs for this.

aanm reviewed (changes requested):
Some of my comments can't be done as a follow up, we should merge this ASAP. :)

raybejjani commented on pkg/ip/ip.go:
```diff
@@ -714,3 +714,31 @@ func partitionCIDR(targetCIDR net.IPNet, excludeCIDR net.IPNet) ([]*net.IPNet, [
+// The slice is manipulated in-place destructively.
+//
+// 1- Sort the slice by comparing the IPs as bytes
+// 2- For every unseen unique IP in the sorted slice, move it to the start of
```
oops, I just realised that this sentence was always wrong! If you end up touching the PR again, and have time, could you replace `move it to the start of` with `move it to the end of`? I can also do it later, no problem.

ianvernon commented on pkg/fqdn/cache.go:
```diff
@@ -15,11 +15,10 @@
 package fqdn
```
The commit message is wrong.

ianvernon commented on pkg/policy/groups/actions.go:
```diff
@@ -135,24 +135,30 @@ func addDerivativeCNP(cnp *cilium_v2.CiliumNetworkPolicy) error {
 		}
-		scopedLog.WithError(err).Error("Cannot create derivative")
-		statusErr := updateDerivativeStatus(cnp, derivativeCNP.ObjectMeta.Name, err)
+		scopedLog.WithError(derivativeErr).Error("Cannot create derivative rule. Installing block all rule.")
```
Nit: instead of "block all" say "deny-all" - this is the preferred vocabulary for such rules, per our documentation.

ianvernon commented on pkg/policy/groups/actions.go:
```diff
@@ -135,24 +135,30 @@ func addDerivativeCNP(cnp *cilium_v2.CiliumNetworkPolicy) error {
+		statusErr := updateDerivativeStatus(cnp, derivativeCNP.ObjectMeta.Name, derivativeErr)
 		if statusErr != nil {
-			log.WithError(err).Error("Cannot update CNP status on invalid derivative")
+			scopedLog.WithError(statusErr).Error("Cannot update CNP status on invalid derivative")
```
"cannot update CNP status for derivative policy"? "invalid derivative" is a bit vague.
